spec/coupon_code_spec.rb

```diff
@@ -1,56 +1,76 @@
+# frozen_string_literal: true
+
 require 'spec_helper'
 
 RSpec.describe CouponCode do
-
   describe '.generate' do
     subject { described_class.generate }
+
     it { is_expected.not_to be_nil }
-    it { is_expected.to match(/^[0-9A-Z-]+$/) }
+    it { is_expected.to match(/^[2-9A-Z-]+$/) }
     it { is_expected.to match(/^\w{4}-\w{4}-\w{4}$/) }
     it 'generates a different code' do
       code2 = described_class.generate
       is_expected.not_to eq(code2)
     end
 
-    context '2 parts' do
+    context 'when 2 parts' do
       subject { described_class.generate(parts: 2) }
       it { is_expected.to match(/^\w{4}-\w{4}$/) }
     end
+
+    context 'when custom part length' do
+      subject { described_class.generate(part_length: 5) }
+
+      it { is_expected.to match(/^[2-9A-Z-]+$/) }
+      it { is_expected.to match(/^\w{5}-\w{5}-\w{5}$/) }
+    end
+
+    context 'with curse word characters' do
+      before do
+        # Sequence with a mix of curse word separators and other characters
+        # Checkdigit also can be curse word separator, so current part will be regenerated in this case
+        allow(CouponCode).to receive(:random_symbol).and_return(*'ABUAFUBCA2FUCP'.chars)
+      end
+
+      it 'avoids generating codes with offensive word excluded characters' do
+        disallowed_pairs = CouponCode::CURSE_WORD_SEPARATORS.product(CouponCode::CURSE_WORD_SEPARATORS).map(&:join)
+        disallowed_regex = Regexp.union(disallowed_pairs)
+        is_expected.not_to match(disallowed_regex)
+      end
+    end
   end
 
   describe '.validate' do
-
     it 'validates a good code' do
-      expect(described_class.validate('1K7Q-CTFM-LMTC')).to eq('1K7Q-CTFM-LMTC')
+      expect(described_class.validate('9JRW-QTJ7-3U5G')).to eq('9JRW-QTJ7-3U5G')
     end
 
     it 'validates and returns the code in uppercase letters' do
-      expect(described_class.validate('1K7Q-ctfm-LMTC')).to eq('1K7Q-CTFM-LMTC')
+      expect(described_class.validate('9jrw-qtj7-3u5g')).to eq('9JRW-QTJ7-3U5G')
     end
 
     it 'returns nil for an invalid code' do
-      expect(described_class.validate('1K7Q-CTFM')).to be_nil
+      expect(described_class.validate('9jrw-qtj7')).to be_nil
     end
 
     it 'handles invalid characters' do
-      expect(described_class.validate('OK7Q-CTFM-LMTC')).to be_nil
+      expect(described_class.validate('9JRF-QTJ7-3U5G')).to be_nil
     end
```

Comment:
I think we can use the same as for `context 'valid cases'`, list of invalid codes

Reply:
I'm not sure. Here we mostly check digit validation but agree it's not so clear from description.